### ISSUE TITLE
[Performance] vLLM warmup does not cover C128A extra_topk=384 on SM120 sparse-MLA decode, causing SparseMlaDecodeV3Runner tactic=-1 fallback

### ISSUE THREAD

**Issue by wuhangxian** [vllm-project/vllm#53601]
# [Performance] vLLM warmup does not cover C128A `extra_topk=384` on SM120 sparse-MLA decode, causing `SparseMlaDecodeV3Runner tactic=-1` fallback

## Summary

vLLM startup warmup for DeepSeek-V4-Flash with SM120 (RTX PRO 5000 Blackwell) sparse-MLA decode only generates FlashInfer autotune cache entries for `extra_topk={0, 128, 512}`. At runtime, when a request\'s context length exceeds 32,895 tokens, the C128A compressed cache selects `extra_topk=384`. Since this exact width is missing from the autotune cache, every decode step falls back to `SparseMlaDecodeV3Runner tactic=-1`, degrading performance.

The root cause is in the warmup path: `_SPARSE_MLA_MIXED_WARMUP_TOKENS = 16` (in `flashinfer_sparse_mla_warmup.py`) creates a mixed batch where decode tokens have `seq_lens=[1]`, which maps to `active_topk_width=128`. The warmup never simulates long-sequence decode, so `extra_topk=384` is never exercised.

## Environment

| Item | Value |
| --- | --- |
| GPU | 4 × NVIDIA RTX PRO 5000 72GB Blackwell (SM120), TP4 |
| NVIDIA Driver | 580.126.20 |
| vLLM | `0.1.dev20005+g75c2eef60` (custom build based on current main) |
| flashinfer-python | `0.6.18.dev20260811` |
| flashinfer-cubin | `0.6.18.dev20260811` |
| flashinfer-jit-cache | `0.6.18.dev20260811+cu130` |
| PyTorch | `2.13.0+cu130` |
| CUDA | 13.0 |
| Model | DeepSeek-V4-Flash (FP8 KV cache, speculative with DSpark, 5 spec tokens) |
| Container image | `hai-deepseek-v4-flash-vllm:latest` (`sha256:113c108a11f3...4086cee`) |

### Launch command

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 \\
python3 -m vllm.entrypoints.openai.api_server \\
  --model /model \\
  --host 0.0.0.0 \\
  --port 30013 \\
  --trust-remote-code \\
  --served-model-name DeepSeek-V4-Flash \\
  --kv-cache-dtype fp8 \\
  --gpu-memory-utilization 0.8 \\
  --speculative-config \'{"method":"dspark","num_speculative_tokens":5}\' \\
  --max-model-len 33792 \\
  --max-num-batched-tokens 8448 \\
  --max-num-seqs 32 \\
  --tensor-parallel-size 4 \\
  --seed 0 \\
  --no-enable-prefix-caching
```

Note: `enable_flashinfer_autotune=True` is set in the kernel config by default for this model.

### Observed behavior

During startup, the FlashInfer autotune warmup generates **24 cache entries** with `extra_topk ∈ {0, 128, 512}` only — **zero entries for `extra_topk=384`**.

During CUDA Graph capture, **944 `tactic=-1` fallback warnings** are emitted across all 4 TP workers, covering 4 unique batch shapes:

| Batch (T) | extra_topk | T bucket | Warning count (per shape, ×4 TP) |
| ---: | ---: | ---: | ---: |
| 6 | 384 | 8 | 236 |
| 12 | 384 | 16 | 236 |
| 24 | 384 | 32 | 236 |
| 60 | 384 | 64 | 236 |

**Total: 944 warnings** (236 × 4 unique shapes, each emitted by all 4 TP workers).

Example warning:
```
WARNING - autotuner.py:1513 - flashinfer.jit: [AutoTuner]: No tuned config covers
sparse_mla_sm120_decode_dsv4 input_shapes=((60, 16, 512), (60, 128), (60, 16, 8, 512),
(60, 16, 8), (60, 16, 512), (60, 16), (60,), (16,), (60, 384), (60,));
falling back to runner=SparseMlaDecodeV3Runner tactic=-1.  This shape is outside the
tuning bucket range -- expand tuning_buckets / max_num_tokens during the next tuning
pass to avoid this perf cliff.
```

### Cache analysis

The autotune cache file contains 24 entries, all with `extra_topk ∈ {0, 128, 512}`:

| extra_topk | Records per T bucket {1,4,8,16,32,64} | Total |
| ---: | ---: | ---: |
| 0 | 1 each | 6 |
| 128 | 1 each | 6 |
| 512 | 1 each (varying tactics) | 6 |
| **384** | **0** | **0** |

Cache SHA-256: `38bf5e3d0b960250eff60e9ea5ea77fe7d29b989c88f1d3313cc6daadd805265`

### Fail-closed request validation

Three requests were sent to the running server. All completed with exact requested output token counts. The cache file was unchanged before and after (SHA-256 verified):

| Input tokens | Output tokens | Elapsed | Finish reason |
| ---: | ---: | ---: | --- |
| 8193 | 512 | 1.92 s | length |
| 16513 | 64 | 1.59 s | length |
| 32897 | 64 | 3.13 s | length |

Cache SHA-256 before: `38bf5e3d0b960250eff60e9ea5ea77fe7d29b989c88f1d3313cc6daadd805265`
Cache SHA-256 after:  `38bf5e3d0b960250eff60e9ea5ea77fe7d29b989c88f1d3313cc6daadd805265`
Cache unchanged: **Yes**

### 1. C128A runtime width reaches 384

In `vllm/models/deepseek_v4/sparse_mla.py`, the C128A compressed cache width is computed as:

```python
# sparse_mla.py:166-170
c128a_max_compressed = cdiv(max_model_len, compress_ratio)       # cdiv(33792, 128) = 264
c128a_max_compressed = cdiv(c128a_max_compressed, _C128A_TOPK_ALIGNMENT) * _C128A_TOPK_ALIGNMENT
                                                                    # cdiv(264, 128) * 128 = 384
```

At runtime, the active width is dynamically selected (`sparse_mla.py:261-265`):

```python
active_topk_width = min(
    max(
        triton.next_power_of_2(max(cm.max_seq_len // self.compress_ratio, 1)),
        _C128A_TOPK_ALIGNMENT,   # 128
    ),
    self.c128a_max_compressed,    # 384
)
```

For `max_model_len=33792`, the static reachable widths are:

| active max_seq_len | extra_topk |
| ---: | ---: |
| 1–16,511 | 128 |
| 16,512–32,895 | 256 |
| 32,896–33,792 | 384 |

### 2. FlashInfer cache key includes exact `extra_topk`

FlashInfer\'s sparse-MLA tuning key includes the exact `extra_topk` value. A width-512 tactic cannot cover width-384. The T dimension uses ceiling buckets `{1, 4, 8, 16, 32, 64}` (e.g. T=60 maps to bucket 64), so the miss is specifically on the width dimension.

### 3. Warmup never exercises long-sequence decode

The warmup path in `vllm/model_executor/warmup/flashinfer_sparse_mla_warmup.py` uses:

```python
_SPARSE_MLA_MIXED_WARMUP_TOKENS = 16
```

This creates a mixed batch (`create_mixed_batch=True`) in `_dummy_run` where decode tokens have `seq_lens=[1]` (`gpu_model_runner.py:6131`):

```python
seq_lens = torch.tensor([1] * num_decode_tokens + [num_prefill_tokens + 1], ...)
```

With `seq_lens=1`, `active_topk_width = min(max(next_power_of_2(1//128), 128), 384) = 128`.

The warmup **never simulates a decode request with `seq_len > 32895`**, so `extra_topk=384` is never autotuned.

## Expected behavior

- Startup autotune covers every runtime-reachable C128A width, including 384 (and 256).
- A production shape does not use `tactic=-1` only because warmup selected a different sequence-length bucket.
- Existing cache entries survive successful tuning.
- Partial tuning failure preserves the baseline cache and cannot deadlock TP followers.
- Repeated startup with a complete cache is idempotent.

## Proposed fix

The warmup should enumerate runtime-reachable C128A widths and exercise each through the model warmup / autotune path. Concretely, in `flashinfer_sparse_mla_warmup.py` or `gpu_model_runner.py`, the dummy run should include at least one decode request with `seq_len` large enough to trigger `extra_topk=384` (i.e. `seq_len ≥ 32896` for `max_model_len=33792`).

Alternatively, if FlashInfer exposes a supported public shape-autotune entrypoint, vLLM could call it on the TP leader to pre-populate the missing width.

## Additional context

- This issue is separate from #53257 and #53436.
- The same defect was independently reproduced on vLLM current main (`a047e2543`) with FlashInfer 0.6.17 (pinned dependency), confirming it is not version-specific to 0.6.18.dev.
- In the motivating environment, adding real width-384 tactics (manual cache intervention) removed all fallback warnings but did not make target logits bit-reproducible. A linked PR must not claim to fix greedy-output or speculative-acceptance non-determinism.

## Before submitting a new issue

- [x] Searched open and closed vLLM issues and PRs for `extra_topk`, `384`, `sparse MLA`, `FlashInfer`, `autotune`, and `warmup`.
- [x] Found related reports, but no issue or PR covering the missing C128A production-width tuning keys.
- [x] Reproduced the scoped width-384 defect on the production environment (FlashInfer 0.6.18.dev20260811) and independently on current main + FlashInfer 0.6.17.

**Comment by jeejeelee** [comment]
duplicate of #53600
